kafka-webview-ui/src/assembly/distribution/start.sh

```diff
@@ -1,23 +1,15 @@
-#!/bin/bash
+#!/bin/sh
 
+# Temporary save current directory
 CWD=`pwd`
 
-## Change to local directory
+# Change to local directory
 cd "${0%/*}"
 
-# Define empty options as defaults if none set
```

Comment:
changing these environment names will break backwards compatibility yea?

Reply:
I tested it manually and I don't see problems.
Unfortunately I was not able to test in Docker, because the build fail. It's not able to download from the artifact repo :(
As soon as I can, I will attach logs for this.

Reply:
I think the issue is if someone has built a deployment process around using these environment variables, removing/renaming them would be a breaking change for them as they would need to update their deployment to use the new variable names.

I'd prefer to maintain backwards compatibility where possible to avoid causing version upgrade pains/difficulties for people.

Reply:
I know what you mean, but sometimes it's unavoidable to have some kind of disruptions after a while between two different versions. The main example is Java 8 to 9. Some disruptive changes were unavoidable and better for the future of Java in general.
Moreover if the documentation is written well, I think everyone can figure out easily what to change and how.

Analysing specifically these changes:
- `HEAP_OPTS` refers just to some specific Java configurations about the memory, what about other configurations? Like `-noverify -server -XX:TieredStopAtLevel=1` these should not be included as this would mean misusing this specific environment variable. A variable called `JVM_OPTS` and `JAVA_OPTS` are more general purpose and usable in many different flexible ways.
- `LOG_OPTS` can be restored back, but it's kind of the same as `HEAP_OPTS`.
- The usage of `SPRING_CONFIG_LOCATION` is completely the same, I removed this as I already included everything in the two default files `/resources/application.yaml` and `/resources/application-dev.yaml`. This has 100% backward compatibility.

Moreover the definition of empty environment variables like
```
if [[ -z "$HEAP_OPTS" ]]; then
    export HEAP_OPTS=""
fi
if [[ -z "$LOG_OPTS" ]]; then
    export LOG_OPTS=""
fi
```
is useless and equivalent to not defining them, so they can be removed without backward compatibility problems anyway.

I missed to share with you a bigger vision on my series of PRs. In the last one, improving Dockerfile, I would remove the usage of `start.sh` in the Dockerfile, in favour to something like following:
```
FROM openjdk:8-jre-alpine

...

ENTRYPOINT exec java $JVM_OPTS $LOG_OPTS $JAVA_OPTS -jar kafka-webview-ui-*.jar
```
`start.sh` does nothing special/specific, so this could be avoided in favour to an easier Dockerfile and no startup scripts, replacing some useless complexity.

Reply:
In regards to the start.sh changes.  It may be better to have a check if the to be deprecated environment variables are used, and display an appropriate warning and migration information.  Until removal, it should be easy to merge the previous variable into the new one.

I see no issues with the Dockerfile avoiding the `start.sh` entirely as you've laid above.

Reply:
> In regards to the start.sh changes. It may be better to have a check if the to be deprecated environment variables are used, and display an appropriate warning and migration information. Until removal, it should be easy to merge the previous variable into the new one.

This is a good idea! We can still have a `start.sh` checking the usage of deprecated environment variables, explaining how to move to new ones. That script would be considered deprecated as well and replaced by a specific section in the documentation. Something like `How to migrate from v2.5.0 to 2.x.y`?
I will adapt the `start.sh` script to check deprecated environment variables asap.

Reply:
Keep in mind I don't think `start.sh` will be removed.  Not everyone deploys the application via Docker, so having a dead simple script to start up the app makes sense for those environments.

Reply:
Got it.
Do you think that the Makefile (to be introduced by PR #200 ) is enough as alternative to `start.sh`?

Reply:
When you have time, can you have a look at following logs about Docker build. It fails locally...

```

docker build -t sourcelaborg/kafka-webview .
[+] Building 62.7s (7/8)
 => [internal] load build definition from Dockerfile                                                                                                                                                                                   0.0s
 => => transferring dockerfile: 1.53kB                                                                                                                                                                                                 0.0s
 => [internal] load .dockerignore                                                                                                                                                                                                      0.0s
 => => transferring context: 2B                                                                                                                                                                                                        0.0s
 => [internal] load metadata for docker.io/library/openjdk:8-jre-alpine                                                                                                                                                                2.2s
 => [1/5] FROM docker.io/library/openjdk:8-jre-alpine@sha256:f362b165b870ef129cbe730f29065ff37399c0aa8bcab3e44b51c302938c9193                                                                                                         55.5s
 => => resolve docker.io/library/openjdk:8-jre-alpine@sha256:f362b165b870ef129cbe730f29065ff37399c0aa8bcab3e44b51c302938c9193                                                                                                          0.0s
 => => sha256:f7a292bbb70c4ce57f7704cc03eb09e299de9da19013b084f138154421918cb4 3.42kB / 3.42kB                                                                                                                                         0.0s
 => => sha256:e7c96db7181be991f19a9fb6975cdbbd73c65f4a2681348e63a141a2192a5f10 2.76MB / 2.76MB                                                                                                                                         7.4s
 => => sha256:f910a506b6cb1dbec766725d70356f695ae2bf2bea6224dbe8c7c6ad4f3664a2 238B / 238B                                                                                                                                             2.4s
 => => sha256:b6abafe80f63b02535fc111df2ed6b3c728469679ab654e03e482b6f347c9639 54.94MB / 54.94MB                                                                                                                                      54.6s
 => => sha256:f362b165b870ef129cbe730f29065ff37399c0aa8bcab3e44b51c302938c9193 1.64kB / 1.64kB                                                                                                                                         0.0s
 => => sha256:b2ad93b079b1495488cc01375de799c402d45086015a120c105ea00e1be0fd52 947B / 947B                                                                                                                                             0.0s
 => => extracting sha256:e7c96db7181be991f19a9fb6975cdbbd73c65f4a2681348e63a141a2192a5f10                                                                                                                                              0.1s
 => => extracting sha256:f910a506b6cb1dbec766725d70356f695ae2bf2bea6224dbe8c7c6ad4f3664a2                                                                                                                                              0.0s
 => => extracting sha256:b6abafe80f63b02535fc111df2ed6b3c728469679ab654e03e482b6f347c9639                                                                                                                                              0.6s
 => [2/5] RUN mkdir -p /app &&     mkdir -p /app/logs &&     mkdir -p /app/data &&     apk add --update bash curl &&     rm -rf /var/cache/apk/*                                                                                       4.0s
 => [3/5] WORKDIR /app                                                                                                                                                                                                                 0.0s
 => ERROR [4/5] RUN curl -fSL -o /tmp/kafka-webview-ui-bin.zip https://oss.sonatype.org/service/local/repositories/orgsourcelab-1031/content/org/sourcelab/kafka-webview-ui/2.5.0/kafka-webview-ui-2.5.0-bin.zip                       0.9s
------
 > [4/5] RUN curl -fSL -o /tmp/kafka-webview-ui-bin.zip https://oss.sonatype.org/service/local/repositories/orgsourcelab-1031/content/org/sourcelab/kafka-webview-ui/2.5.0/kafka-webview-ui-2.5.0-bin.zip:
#7 0.213   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
#7 0.213                                  Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
#7 0.863 curl: (22) The requested URL returned error: 404
------
failed to solve with frontend dockerfile.v0: failed to build LLB: executor failed running [/bin/sh -c curl -fSL -o /tmp/kafka-webview-ui-bin.zip https://oss.sonatype.org/service/local/repositories/orgsourcelab-1031/content/org/sourcelab/kafka-webview-ui/${WEBVIEW_VER}/kafka-webview-ui-${WEBVIEW_VER}-bin.zip]: runc did not terminate sucessfully
```

I'm using Docker Edge for macOS, but I tried with Docker Stable as well.

Reply:
> Got it.
> Do you think that the Makefile (to be introduced by PR #200 ) is enough as alternative to `start.sh`?

Probably doesn't make sense to package a Makefile with a binary distribution that's already compiled.

RE: Docker failing to find the archive from sonatype... I'm not sure.  It looks like all of the releases were removed from their maven repository, I'll need to look into why.

You should be able to swap out the URL for the copy hosted on github here: https://github.com/SourceLabOrg/kafka-webview/releases/download/v2.5.0/kafka-webview-ui-2.5.0-bin.zip

Reply:
> Probably doesn't make sense to package a Makefile with a binary distribution that's already compiled.

Ah ok, you mean really deploy just the `jar` together with the `script.sh` to be deployed on prem. Ok got it. I will think of something like that ;)

> You should be able to swap out the URL for the copy hosted on github here: https://github.com/SourceLabOrg/kafka-webview/releases/download/v2.5.0/kafka-webview-ui-2.5.0-bin.zip

I will prepare some different Dockerfile to have more options to build the project. For example a `Dockerfile.local` to take the jar from local, in order to be able to use a very custom version of Kafka-WebView and a `Dockerfile.release` to be used for an official release.

Reply:
Hi @Crim, sorry to let you wait, Xmas time here around.

I made the changes we discussed on the `start.sh`, please have a look at it!
After my changes in the last PR of this series, I would like to discuss how to deal with the start scripts. I don't understand why they are inside src/assembly for example.

I will put all changes to the Dockerfile in the last PR of this series.

Reply:
So after reviewing this file, I'm not sure it works like you might expect it to.  I've modified line 25 as shown below to show what command would be run.

```
# Start application
echo java $JVM_OPTS $MEM_OPTS $JAVA_OPTS -jar kafka-webview-ui-*.jar
```

As written its impossible to set the new environment variables and have them be used.  Example:

```
SMBP:tmp $ export JAVA_OPTS="-Xms1G -Xmx5G"
SMBP:tmp $ ./test.sh 
java -noverify -server -XX:TieredStopAtLevel=1 -Xms2G -Xmx2G -XX:MaxMetaspaceSize=300M -jar kafka-webview-ui-*.jar
```

I believe you'd need the if not defined checks, then use your default values, instead of just hardcoding the environment variables?

Reply:
@Crim I'm really sorry! I made a big mistake.
I just pushed a new version of the script. I tested it more extensively with `zsh` and `bash`, it works as expected now. Please test it and let me know.

Big changes:
- removed the folder change, it's not necessary if we take care of find the directory of the script and put it in front of the jar file name
- removed `exec` from java execution, not necessary as java command is blocker anyway